[Feature] 텍스트 오브젝트 추가 구현

eemdeeks opened the pull request:
## 🌁 Background
<!-- 해당 PR을 작성하게 된 이유를 적어주세요. -->
텍스트 오브젝트를  추가하는 로직과 뷰를 구현했습니다.

## 📱 Screenshot
<!-- 스크린샷이나 동영상을 첨부해주세요. -->
| iPhone SE3 | iPhone 13 mini | iPhone 16 Pro |
|--------|--------|--------|
|![Simulator Screen Recording - iPhone SE (3rd generation) - 2024-11-19 at 13 19 03](https://github.com/user-attachments/assets/0863c523-6715-44b2-a22a-6e29f0772111)|![Simulator Screen Recording - iPhone 13 mini - 2024-11-19 at 13 22 59](https://github.com/user-attachments/assets/52c60289-732e-40b3-9d24-da1c7e70dbb7)|![Simulator Screen Recording - iPhone 16 Pro - 2024-11-19 at 13 24 56](https://github.com/user-attachments/assets/d3e54c66-a463-4420-8826-336b702b6e55)|


## 👩‍💻 Contents
<!-- 작업 내용을 적어주세요 -->
### 텍스트 오브젝트를 생성하는 로직을 구현했습니다
- 텍스트 오브젝트를 생성하기 위해 현재 화면이 보고있는 offset값을 가져와 처리하도록 해 주었습니다.

### 텍스트 오브젝트 추가 플로우
- 텍스트 오브젝트는 우선, 추가하여 빈 텍스트오브젝트를 추가하고, 전송 해 줍니다.
- 이후 수정 에픽 단계에서 오브젝트의 내용이 변할 경우 전송해주는 로직을 구현 할 계획입니다.

## ✅ Testing
<!-- 테스트 방법을 적어주세요 -->
- 테스트 목적과 상황
텍스트 오브젝트가 정상적인 위치에 생성되는지 유즈케이스를 테스트

- 시나리오 진행에 필요한 값
1. 화면의 크기
2. 지금 보고있는 스크롤뷰의 오프셋값

- 시나리오 진행에 필요한 조건
usecase 생성시 텍스트 오브젝트의 기본 크기를 설정하는 값을 넣어주어야 한다.

- 시나리오 완료 시 보장하는 결과
지금 보고있는 화면의 중상위 부근에 텍스트오브젝트가 생성되어야 한다.

## 📝 Review Note
<!-- PR과정에서 든 생각이나 개선할 내용이 있다면 적어주세요. -->
### TextFieldDelegate를 이용하는 코드를 어디에 넣어야할지 고민입니다.
TextObjectView 내부에 TextField를 갖고있는데요, 만약 뷰컨트롤러에서 delegate를 주입해준다면, 생성시에 다른 WhiteboardObjectView와는 다르게 팩토리에서 delgate를 받아서 뱉어주거나, 따로 주입시켜주기위해 TextObjectView로 형변환을 해서 주입해주어야합니다...

그래서 우선은 TextObjectView자체에서 delegate채택후 주입시켜 주고 있습니다.

다만, 이렇게 될경우에는 delegate 메소드 내부에 들어가는 코드를 어떻게 처리해야할지 고민입니다..

## 📣 Related Issue
<!-- 관련 이슈를 적어주세요. -->
- close #7 


## 📬 Reference
<!-- 참고한 코드의 출처를 작성해주세요 -->

eemdeeks committed:
feature: 텍스트오브젝트생성 유즈케이스 설계(#7)

eemdeeks committed:
feature: 텍스트오브젝트 추가 유즈케이스 구현(#7)

eemdeeks committed:
feature: 텍스트오브젝트 추가 기능 뷰모델에 추가(#7)

eemdeeks committed:
feature: 텍스트오브젝트 추가 뷰 구현(#7)

eemdeeks committed:
feature: 텍스트오브젝트 포커스 기능 구현(#7)

eemdeeks committed:
feature: 텍스트오브젝트 추가 테스트코드(#7)

taipaise commented on Domain/Domain/Sources/Interface/UseCase/TextObjectUseCaseInterface.swift:
```diff
@@ -0,0 +1,19 @@
+    ///   - scrollViewOffset: 현재 스크롤뷰가 보고있는 위치
+    ///   - viewSize: 현재 뷰의 크기
+    /// - Returns: 현재 화면 중앙에 위치한 TextObject를 반환
+    func addText(scrollViewOffset: CGPoint, viewSize: CGSize) -> TextObject
```
useCase의 파라미터 이름을 보았을때, `UseCase`가 `ScrollView`를 알고 있다고 느껴집니다!
View에 종속적인 네이밍이라고 생각하는데요, `point` 와 같은 네이밍은 어떨까요?
`viewSize`도 `size` 와 같은 네이밍은 어떤가요??

eemdeeks replied:
그렇게 느껴지네요!!
수정해보겠습니다! 구욷

taipaise commented on Domain/DomainTests/TextObjectUseCaseTests.swift:
비정상적인 입력이 들어왔을때도 일단은 TextObject 생성이 되는것 같습니다! 생성을 막지 않고 정상 case와 같이 동작하도록 하신 이유가 있을까요?

eemdeeks replied:
뷰에 대한 데이터만 들어온다고 생각했을 때,
저희가 생각하지 못한 뷰의 값들이 들어올 수도 있다고 생각했습니다!

사람이 생각하지 못한 실제 뷰에서도 동작을 하도록 했는데요, 해당 경우를 관리할 수 있도록 해주는 것이 더 좋을까요?

taipaise replied:
저도 아직 다른 ObjectView에서 적절하지 않은 입력에 대한 처리를 안해서,, 추후 처리 로직을 함께 추가해도 좋을 것 같습니다!

eemdeeks replied:
확인했습니다!

choijungp commented on Presentation/Presentation/Sources/Whiteboard/View/ObjectView/TextObjectView.swift:
```diff
@@ -9,13 +9,14 @@ import Domain
 
 final class TextObjectView: WhiteboardObjectView {
-    let textField: UITextField = {
+    private lazy var textField: UITextField = {
         let textField = UITextField()
         textField.placeholder = "Hello AirplaIN"
+        textField.delegate = self
         return textField
     }()
```
요기서 `lazy var` 사용하는 방법 대신     
private let으로 선언 후에 `init`이나 `configureAttribute `함수에서 textField의 delegate를 잡아주는건 어떨까요 ?? 

방어적인 코드를 사용하기 위해 lazy var를 지양하기로 했던 것이 생각나서 말씀드려봅니다. .

eemdeeks replied:
저희는 TextView를 만들어 반환하는 Factory를 사용하고 있습니다.

위에서 했던 고민과 마찬가지로 
Factory를 사용해 뷰를 만들어 반환하는 메소드가 추상화가 되어있는데요,
이 때 TextView에서만 delegate를 사용하기에 해당 방법(configure와 같이)을 사용한다면 뷰 펙토리 생성 후 따로 as를 통해 형변환 후 해주어야 했습니다.

그래서 우선 이렇게 구현을 했는데, 더 좋은 방법을 따져봐야할 것 같습니다..

taipaise replied:
타입 캐스팅 때문에 고민이시라면, 아래와 같은 코드는 어떨까요?
```swift
 private let textField: UITextField = {
        let textField = UITextField()
        textField.placeholder = "Hello AirplaIN"
        return textField
    }()

    private let textObject: TextObject

    init(textObject: TextObject) {
        self.textObject = textObject
        super.init(whiteboardObject: textObject)
        configureAttribute()
        configureLayout()
    }

    required init?(coder: NSCoder) {
        self.textObject = TextObject(
            id: UUID(),
            position: .zero,
            size: .zero,
            text: "")
        super.init(coder: coder)
    }

    override func becomeFirstResponder() -> Bool {
        textField.becomeFirstResponder()
    }

    private func configureAttribute() {
        textField.delegate = self
    }
```

eemdeeks replied:
해당 방법으로 lazy에 대한 해결이 가능 할 것 같습니다!!
그렇다면 현재 view가 delegate를 채택하고 넣어주는 것은 현상유지로 가도 괜찮을까요?

taipaise replied:
넵 좋습니다!! 다만 텍스트 뷰에서 수정사항에 대해 VC에 알리는 delegate가 추가될 수 있을 것 같은데요..!
저희가 말씀하신 것 처럼 factory에서 view를 만들어 WhiteboardObjectView로 반환하고 있기 때문에, TextObjectViewDelegate 가 추가되었을 때 delegate 처리가 애매해질 수는 있을것 같습니다..!
factory에서 해당 delegate를 채택한 VC를 weak 프로퍼티로 가지고 있으면 해결 할 수 있을 것 같기도 하네요!!

eemdeeks replied:
좋은 방법인 것 같습니다!!
그렇다면 팩토리를 추상화한 인터페이스가 갖고있어야 할 것 같은데, 
여러 방법과 추상화의 의미를 좀 더 생각해 보아야할 것 같습니다 :(

taipaise replied:
앗,, 거기까지는 생각하지 못했네요.. 생각보다 복잡하군요!!!

choijungp commented on Presentation/Presentation/Sources/Whiteboard/View/ObjectView/TextObjectView.swift:
```diff
@@ -33,9 +34,30 @@ final class TextObjectView: WhiteboardObjectView {
 }
+
+extension TextObjectView: UITextFieldDelegate {
+    func textField(
+        _ textField: UITextField,
+        shouldChangeCharactersIn range: NSRange,
+        replacementString string: String
+    ) -> Bool {
+        let maxLength = 15
+        guard let text = textField.text else { return true }
+        let newlength = text.count + string.count - range.length
+        return newlength < maxLength
+    }
```
text 오브젝트의 글자수를 15자로 제한한 이유가 있을까요 ??

eemdeeks replied:
우선 임시로 제한 해 두었습니다.
텍스트 필드 글자 수 제한을 하지 않는다면, 텍스트 필드의 길이에 따라 글이 숨겨지는 현상을 회피하기 위해 해두었습니다..

더 좋은 방법이 있다면 추천해주시면 시도해보겠습니다!

choijungp commented on Presentation/Presentation/Sources/Whiteboard/View/ObjectView/TextObjectView.swift:
```diff
@@ -33,9 +34,30 @@ final class TextObjectView: WhiteboardObjectView {
+        return newlength < maxLength
+    }
+
+    func textFieldShouldReturn(_ textField: UITextField) -> Bool {
+        textField.resignFirstResponder()
+    }
```
허얼 ~ 저 프로필 뷰에 이 함수 안썼던 것 같아요 ! 버그 수정할 때 까먹지 않고 추가하겠슴둥 !!!

eemdeeks replied:
좋습니다 :)

choijungp commented on Presentation/Presentation/Sources/Whiteboard/ViewModel/WhiteboardViewModel.swift:
```diff
@@ -8,13 +8,14 @@ import Combine
 import Foundation
 
-final class WhiteboardViewModel: ViewModel {
+public final class WhiteboardViewModel: ViewModel {
```
이것이 public이여야 하는 이유가 있나요 ??
저는 AirplaIN 앱 모듈에서 첫 진입화면에 사용되어야 하는 것들만 Public이 된다고 생각했습니다. !! 
(사실 근데 그 이후에 의존성 주입까지 생각하면 public이 될지도 모르겠지만 ... 당장은 public이여야 한다는 확신이 없습니다 !!)

eemdeeks replied:
동의합니다!
의존성 주입을 해야할 때 아마 public이 되어야할 것 같은데요,
이미 외부에서 사용한다면 모듈들의 의존성을 관리해 주고 있기 때문에 당장에는 괜찮지 않을까요..?

지극히 개인적인 생각입니다..

taipaise replied:
두 의견 모두 동의합니다~!~!
앱 모듈에서 의존성을 만들어줘야하니 결국에는 public으로 선언이 되겠지만,, 내부 프로퍼티나 메서드들은 internal 또는 private으로 선언되어 있어 자연스럽게 캡슐화가 되지 않을까 합니다!

eemdeeks replied:
우선 수정을 했습니다!
지금의 PR에서는 필요없는 코드라고 볼 수도 있어, 수정했습니다!
이후에 필요에 의해 수정하면 될 것 같습니다!

choijungp reviewed (approved):
확인 !!!!! 어푸루부 😇 수고하셨습니딴 !!!

근데 TextField delegate 메서드에서 어떤 작업들을 하길래 TextObjectView에서 처리해줘야 하는게 고민인지 여쭤봐도 될까요 ??

taipaise commented on Presentation/Presentation/Sources/Whiteboard/View/ObjectView/TextObjectView.swift:
텍스트 필드를 통해 텍스트를 보여주고 있는 것 같습니다!
1. 아무것도 입력하지 않을 수 있습니다! 이 경우 placeholder 가 표시되고 있습니다.
2. 이미 추가한 오브젝트도 탭하여 텍스트를 수정할 수 있습니다!

https://github.com/user-attachments/assets/5c216421-7fd3-470a-a528-183417f67fa6

위와 같이 동작했을때, 오브젝트 선택기능 구현 시 textObjectView는 선택이 아닌 텍스트 편집을 하게 될 것 같은데 괜찮을까요?? 의도하신 동작이 맞는지 궁금합니다!!

eemdeeks replied:
PR 내용에 적혀 있듯이, 텍스트 오브젝트를 추가하는 플로우를 생각 했을 때 텍스트 오브젝트를 추가하는 경우 바로 텍스트오브젝트를 생성 해 주고, 보내줍니다.
이 후 수정을 통해 텍스트 필드를 편집할 수 있게 의도 했습니다.

말씀해주신 텍스트 오브젝트 뷰 선택이 텍스트 필드가 focus된 경우라고 생각하고 있었습니다!

taipaise replied:
죠습니다!!! 혹시 그렇다면, TextViewObject는 선택 후 이동을 하고 싶다면, 다른 ObjectView들과는 로직이 달라지는 것일까요??

taipaise replied:
추가로, TextObjectView에서 아무 입력하지 않았을 때, placeholder 대신 기본 텍스트를 제공해주는 방법은 어떤가요?!
아니라면 Placeholder를 회색이 아닌, text 처럼 보이게 검은 색으로 제공하는 방법은 별로일까요??

eemdeeks replied:
이동을 할 때에는 크기를 변경할 때(ex. 오른쪽 아래에 있는 버튼 팬제스쳐)와 같이 오른쪽위와 같은 곳에 버튼 팬제스쳐를 생각하고 있었습니다!!

`+`
placeholder와 같은 경우는 개인적으로 구현 한 것이라 다같이 상의 후 결정하면 더 좋을 것 같습니다!!
추천해 주신 방법들도 좋습니다!

taipaise replied:
오브젝트를 탭 했을 때 textfield가 becomeFirstResponder로 설정되어 있어 탭하면 키보드가 바로 올라올 것 같습니다!
키보드 영역 때문에 텍스트 오브젝트를 원하는 위치로 이동하는게 번거롭지는 않을까요?
1번째 탭 -> 선택, 2번째 탭 -> textfiled 수정과 같은 플로우는 구현이 힘들,,겠죠?

eemdeeks replied:
가능 할 수도 있을 것 같습니다!!

이후에 텍스트오브젝트 수정 PR에 고민해 봐도 괜찮을까요?

taipaise replied:
당근 죠습니다🥕🦈

taipaise commented on Domain/Domain/Sources/UseCase/TextObjectUseCase.swift:
```diff
@@ -0,0 +1,28 @@
+        self.textFieldDefaultSize = textFieldDefaultSize
+    }
+
+    public func addText(scrollViewOffset: CGPoint, viewSize: CGSize) -> TextObject {
```
텍스트 오브젝트를 추가할 때, TextObject는 위치와 크기 정보만 가지고 있게 될 것 같습니다.
텍스트 정보를 엔티티에서 가지고 있지 않아도 괜찮을까요?

eemdeeks replied:
PR내용에서 텍스트추가에 대한 플로우를 간략하게 적어 놓았는데요,
구현한 내용 자체는 텍스트오브젝트 추가에 대한 유즈케이스만 작성 했습니다.

위 플로우 대로라면 수정을 했을 경우에 대한 유즈케이스가 추가 될 것으로 계획하고 있습니다!!
그 때 텍스트 정보를 엔티티에서 가지고 있게 될 것 같습니다!
현재 텍스트 정보는 빈 값이어야 한다고 생각했습니다!

taipaise commented on Presentation/Presentation/Sources/Whiteboard/View/WhiteboardViewController.swift:
번거로우시겠지만, text tool을 사용하고 나면, viewModel에게 finishUsingTool라는 input을 전달해 주어야 사용중인 tool이 nil로 바뀝니다.. 
이렇게 보니 좋지 않은 설계같기도 한데요, 추후 같이 개선하면 좋을 것 같습니다!!

eemdeeks replied:
이해 했습니다!
그래서 아직 텍스트필드가 포커싱 중이라면, 해당 텍스트를 선택하고 수정중이라고 생각하여 그대로 두었습니다.
이후에 구현될 Done이라는 버튼이나 키보드의 return버튼을 통해 finishUsingTool을 전달 해 주면 될 것 같습니다!

taipaise commented on Domain/Domain/Sources/UseCase/TextObjectUseCase.swift:
```diff
@@ -0,0 +1,28 @@
+        return TextObject(
+            id: UUID(),
+            position: position,
+            size: textFieldDefaultSize,
```
긴 문자열을 입력했을 때, 문자열이 잘리는 경우는 없을까요??

eemdeeks replied:
그경우를 생각해서 TextObjectView에서 임시로 텍스트필드에 입력할 수 있는 값을 정해 주었습니다!

이후에 상의를 통해서 적절한 defaultSize와 입력할 수 있는 글자 수를 정해보면 될 것 같습니다!

taipaise replied:
goood 길이 처리가 애매할 것 같아서 freeform을 참고 했는데요, 최대 길이를 제한하고 있는 것 같군요!
현재방법 처럼 텍스트 길이 제한이나, view의 최대 길이 제한을 두는 방법 둘 다 좋은 것 같습니다~!~!

eemdeeks replied:
좋습니다!! 다행히도 freeform도 같은 방법을 채택하고 있었군요.. 미리 확인 해봤으면 더 좋았을텐데 다행입니다ㅎㅎ

taipaise reviewed (commented):
딴리 완~~ 고생하셨습니다! 코멘트 확인해주시면 감사하겠습니다

eemdeeks committed:
refactor: 코드리뷰 반영

- Domain에서 뷰에 대한 정보를 알 수도 있다는 암시적인 파라미터 네이밍을 수정
- 현재 필요없는 구현체를 Public으로 되어있던 것 수정

ekrud99 commented on Domain/DomainTests/TextObjectUseCaseTests.swift:
```diff
@@ -0,0 +1,74 @@
+
+    // addText테스트
+    // 비정상적인 입력(0)
+    func testStrangeZeroAddText() {
```
네이밍 관련이라 무시하셔도 되는데..
Strange보다는 Invalid가 어떠신지요..?
`testInvalidInputOfZero`??

eemdeeks replied:
오 좀 더 테스트 이름으로 테스트에 대한 것을 이해할 수 있을 것 같습니다!!

taipaise reviewed (approved):
고생하셨습니다!!! 저는 label로 text 추가를 생각했는데요, 추후 수정까지 생각하면 textField를 통해 구현하는 것이 훨씬 좋다 느껴집니다 bb
왕년에 씽크빅 좀 하셨군요?

ekrud99 reviewed (approved):
고민 많으셨을텐데 고생하셨습니다 ㅎㅎ

eemdeeks commented:
@choijungp 
> 확인 !!!!! 어푸루부 😇 수고하셨습니딴 !!!
> 
> 근데 TextField delegate 메서드에서 어떤 작업들을 하길래 TextObjectView에서 처리해줘야 하는게 고민인지 여쭤봐도 될까요 ??

우선 텍스트필드 딜리게이트를 뷰컨에서 넣어주는지, 뷰에서 처리해줘야하는지 고민이 된 이유가, delegate관련 코멘트 댓글에서 볼 수 있듯이
factory를 통해 만들어 add 해 줄 때, 뷰컨에서 넣어준다면 타입캐스팅이 필요할 것으로 생각했습니다..!
해당 코멘트에서 뷰 내에서 한다면 처리가 가능하게 될 것으로 보이는데, 이 또한 딜리게이트를 뷰가 갖고있어도되는지가 의문입니다..!

eemdeeks committed:
refactor: 쓰레드세이프하지 않은 코드 수정

lazy var를 사용하던 view 프로퍼티 수정

eemdeeks committed:
Merge branch 'develop' into feature/textObject
